### ISSUE TITLE
Memoize isValidZone

### ISSUE THREAD

**Issue by willemdewit** [moment/luxon#479]
Thank you for this great library. Let's make it even better!

I'm currently investigating a performance issue and I see that the `IANAZone#isValidZone` is a bottleneck. This is due to the fact that a function with a try/catch block cannot be optimized by the browser. 
Is it a possibility that the `isValidZone` is memoized so that converting dates between zones becomes a lot faster?

Should I open a PR for this?

**Comment by icambron** [comment]
It's a little surprising to me that `isValidZone` is your bottleneck, not because it isn't slow (it is), but because we cache the zones themselves, so in most scenarios you'd only ever end up checking `isValidZone` once per zone anyway. Can you tell me more about your use case?

**Comment by romamd** [comment]
@icambron Hi! Actually I have the same problem. It looks like Intl.DateTimeFormat has a memory leak. I have run some benchmarks on node 10.7.0 and printing total memory after each cycle:
```
<Test #1> x 1,180 ops/sec ±0.00% (0 runs sampled) memory: 95MB
<Test #1> x 1,913 ops/sec ±0.00% (0 runs sampled) memory: 97MB
<Test #1> x 1,619 ops/sec ±0.00% (0 runs sampled) memory: 106MB
<Test #1> x 1,902 ops/sec ±0.00% (1 run sampled) memory: 109MB
<Test #1> x 1,739 ops/sec ±102.05% (2 runs sampled) memory: 113MB
<Test #1> x 1,825 ops/sec ±19.93% (3 runs sampled) memory: 117MB
<Test #1> x 1,915 ops/sec ±11.10% (4 runs sampled) memory: 122MB
<Test #1> x 1,915 ops/sec ±8.81% (5 runs sampled) memory: 126MB
<Test #1> x 1,753 ops/sec ±7.28% (6 runs sampled) memory: 131MB
<Test #1> x 1,862 ops/sec ±5.94% (7 runs sampled) memory: 135MB
<Test #1> x 1,832 ops/sec ±5.08% (8 runs sampled) memory: 140MB
<Test #1> x 1,866 ops/sec ±4.38% (9 runs sampled) memory: 145MB
<Test #1> x 1,730 ops/sec ±3.91% (10 runs sampled) memory: 149MB
<Test #1> x 1,828 ops/sec ±3.62% (11 runs sampled) memory: 154MB
<Test #1> x 1,923 ops/sec ±3.27% (12 runs sampled) memory: 158MB
<Test #1> x 1,969 ops/sec ±3.27% (12 runs sampled) memory: 163MB
<Test #1> x 1,905 ops/sec ±3.27% (12 runs sampled) memory: 167MB
<Test #1> x 1,865 ops/sec ±3.10% (13 runs sampled) memory: 172MB
<Test #1> x 1,945 ops/sec ±2.88% (14 runs sampled) memory: 176MB
<Test #1> x 1,892 ops/sec ±2.82% (15 runs sampled) memory: 181MB
<Test #1> x 1,916 ops/sec ±2.67% (16 runs sampled) memory: 186MB
<Test #1> x 1,825 ops/sec ±2.56% (17 runs sampled) memory: 191MB
<Test #1> x 1,667 ops/sec ±2.40% (18 runs sampled) memory: 195MB
<Test #1> x 1,873 ops/sec ±2.51% (19 runs sampled) memory: 200MB
<Test #1> x 1,922 ops/sec ±2.39% (20 runs sampled) memory: 205MB
<Test #1> x 1,719 ops/sec ±2.32% (21 runs sampled) memory: 210MB
<Test #1> x 1,924 ops/sec ±2.29% (22 runs sampled) memory: 214MB
<Test #1> x 1,942 ops/sec ±2.23% (23 runs sampled) memory: 219MB
<Test #1> x 1,963 ops/sec ±2.19% (24 runs sampled) memory: 224MB
<Test #1> x 1,833 ops/sec ±2.17% (25 runs sampled) memory: 228MB
<Test #1> x 1,816 ops/sec ±2.08% (26 runs sampled) memory: 233MB
<Test #1> x 1,823 ops/sec ±2.00% (27 runs sampled) memory: 238MB
<Test #1> x 1,916 ops/sec ±1.92% (28 runs sampled) memory: 244MB
<Test #1> x 1,928 ops/sec ±1.88% (29 runs sampled) memory: 248MB
<Test #1> x 1,935 ops/sec ±1.84% (30 runs sampled) memory: 253MB
<Test #1> x 1,963 ops/sec ±1.80% (31 runs sampled) memory: 258MB
<Test #1> x 2,016 ops/sec ±1.72% (32 runs sampled) memory: 262MB
<Test #1> x 1,804 ops/sec ±1.72% (32 runs sampled) memory: 267MB
<Test #1> x 1,788 ops/sec ±1.67% (33 runs sampled) memory: 272MB
<Test #1> x 1,834 ops/sec ±1.63% (34 runs sampled) memory: 277MB
<Test #1> x 1,980 ops/sec ±1.58% (35 runs sampled) memory: 282MB
<Test #1> x 1,815 ops/sec ±1.59% (36 runs sampled) memory: 287MB
<Test #1> x 1,938 ops/sec ±1.54% (37 runs sampled) memory: 292MB
<Test #1> x 1,982 ops/sec ±1.52% (38 runs sampled) memory: 297MB
<Test #1> x 1,845 ops/sec ±1.52% (39 runs sampled) memory: 302MB
<Test #1> x 1,901 ops/sec ±1.48% (40 runs sampled) memory: 307MB
<Test #1> x 1,690 ops/sec ±1.45% (41 runs sampled) memory: 312MB
<Test #1> x 1,718 ops/sec ±1.49% (42 runs sampled) memory: 317MB
<Test #1> x 1,891 ops/sec ±1.49% (43 runs sampled) memory: 321MB
<Test #1> x 1,867 ops/sec ±1.46% (44 runs sampled) memory: 326MB
<Test #1> x 1,869 ops/sec ±1.43% (45 runs sampled) memory: 331MB
<Test #1> x 1,800 ops/sec ±1.40% (46 runs sampled) memory: 338MB
<Test #1> x 1,909 ops/sec ±1.37% (47 runs sampled) memory: 343MB
<Test #1> x 1,986 ops/sec ±1.35% (48 runs sampled) memory: 347MB
<Test #1> x 1,958 ops/sec ±1.35% (49 runs sampled) memory: 352MB
<Test #1> x 1,765 ops/sec ±1.34% (50 runs sampled) memory: 357MB
<Test #1> x 2,038 ops/sec ±1.33% (51 runs sampled) memory: 362MB
<Test #1> x 2,090 ops/sec ±1.33% (51 runs sampled) memory: 367MB
<Test #1> x 2,050 ops/sec ±1.33% (51 runs sampled) memory: 372MB
<Test #1> x 1,908 ops/sec ±1.36% (52 runs sampled) memory: 377MB
<Test #1> x 2,088 ops/sec ±1.34% (53 runs sampled) memory: 383MB
<Test #1> x 2,006 ops/sec ±1.37% (54 runs sampled) memory: 389MB
<Test #1> x 2,082 ops/sec ±1.37% (55 runs sampled) memory: 394MB
<Test #1> x 1,927 ops/sec ±1.40% (56 runs sampled) memory: 399MB
<Test #1> x 2,038 ops/sec ±1.38% (57 runs sampled) memory: 404MB
<Test #1> x 2,027 ops/sec ±1.39% (58 runs sampled) memory: 410MB
<Test #1> x 2,105 ops/sec ±1.39% (59 runs sampled) memory: 415MB
<Test #1> x 2,104 ops/sec ±1.39% (59 runs sampled) memory: 420MB
<Test #1> x 2,109 ops/sec ±1.42% (60 runs sampled) memory: 425MB
<Test #1> x 2,062 ops/sec ±1.44% (61 runs sampled) memory: 430MB
<Test #1> x 1,913 ops/sec ±1.45% (62 runs sampled) memory: 436MB
<Test #1> x 2,055 ops/sec ±1.42% (63 runs sampled) memory: 442MB
<Test #1> x 1,929 ops/sec ±1.43% (64 runs sampled) memory: 447MB
<Test #1> x 2,069 ops/sec ±1.41% (65 runs sampled) memory: 452MB
<Test #1> x 2,024 ops/sec ±1.41% (66 runs sampled) memory: 457MB
<Test #1> x 2,042 ops/sec ±1.41% (67 runs sampled) memory: 462MB
<Test #1> x 2,052 ops/sec ±1.40% (68 runs sampled) memory: 467MB
<Test #1> x 1,954 ops/sec ±1.40% (69 runs sampled) memory: 472MB
<Test #1> x 2,060 ops/sec ±1.38% (70 runs sampled) memory: 477MB
<Test #1> x 1,976 ops/sec ±1.38% (71 runs sampled) memory: 484MB
<Test #1> x 2,079 ops/sec ±1.37% (72 runs sampled) memory: 489MB
<Test #1> x 1,986 ops/sec ±1.37% (73 runs sampled) memory: 494MB
<Test #1> x 1,976 ops/sec ±1.36% (74 runs sampled) memory: 500MB
<Test #1> x 2,045 ops/sec ±1.35% (75 runs sampled) memory: 505MB
<Test #1> x 2,062 ops/sec ±1.34% (76 runs sampled) memory: 510MB
<Test #1> x 2,092 ops/sec ±1.34% (77 runs sampled) memory: 515MB
<Test #1> x 2,063 ops/sec ±1.34% (78 runs sampled) memory: 520MB
<Test #1> x 2,107 ops/sec ±1.34% (79 runs sampled) memory: 525MB
<Test #1> x 2,016 ops/sec ±1.34% (80 runs sampled) memory: 532MB
<Test #1> x 2,097 ops/sec ±1.33% (81 runs sampled) memory: 537MB
<Test #1> x 2,031 ops/sec ±1.33% (82 runs sampled) memory: 542MB
<Test #1> x 2,068 ops/sec ±1.33% (83 runs sampled) memory: 547MB
<Test #1> x 2,010 ops/sec ±1.32% (84 runs sampled) memory: 552MB
<Test #1> x 1,948 ops/sec ±1.31% (85 runs sampled) memory: 557MB
```
As you can see memory is continuously increasing. Another problem is that memory doesn't decrease after that, after 20 minutes memory is still high: 497Mb. On node 8 memory is not that high, but still not good, around 300Mb.
Memoization solves the problem, so I will open a PR.

**Comment by icambron** [comment]
It might make more sense to just memoize the creation of the zone objects themselves? We only check `isValidZone` once...

**Comment by romamd** [comment]
> It might make more sense to just memoize the creation of the zone objects themselves? We only check `isValidZone` once...
Makes sense, but, I am thinking maybe a better solution would be to replace all Intl.DateTimeFormat instantiations with a memoized function. I can see something similar in https://github.com/moment/luxon/blob/master/src/impl/locale.js#L8, I can move it to a separate util and reuse it in whole project. What do you think?

**Comment by icambron** [comment]
Definitely constructing instances of Intl.DateTimeFormat are expensive (there's even a cache for them in the IANAZone file for its other use of Intl.DateTimeFormat) and I'm all for caching them. But the `format` call is a separate matter. Not sure how much of the problem is the construction and how much is the method call.

**Comment by romamd** [comment]
@icambron Ok, for now I can just memoize IANAZone creation as you suggested previously

**Comment by icambron** [comment]
Fixed in #481
